enterprise/cmd/executor/internal/worker/workspace/firecracker.go

```diff
@@ -0,0 +1,228 @@
+	"strings"
+	"syscall"
+
+	"github.com/c2h5oh/datasize"
+
+	"github.com/sourcegraph/sourcegraph/enterprise/cmd/executor/internal/command"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/executor"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// NewFirecrackerWorkspace creates a new workspace for firecracker-based execution.
+// A block device will be created on the host disk, with an ext4 file system. It
+// is exposed through a loopback device. To set up the workspace, this device will
+// be mounted and clone the repo and put script files in it. Then, the executor
+// VM can mount this loopback device. This prevents host file system access.
+func NewFirecrackerWorkspace(
+	ctx context.Context,
+	job executor.Job,
+	diskSpace string,
+	commandRunner command.Runner,
+	commandLogger command.Logger,
+	cloneOpts CloneOptions,
+	operations *command.Operations,
+) (Workspace, error) {
+	blockDeviceFile, tmpMountDir, blockDevice, err := setupLoopDevice(
+		ctx,
+		job.ID,
+		diskSpace,
+		commandLogger,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmount the workspace volume when done, we finished writing to it from the host.
+	defer func() {
+		if err2 := syscall.Unmount(tmpMountDir, 0); err2 != nil {
+			err = errors.Append(err, err2)
+			return
+		}
+		if err2 := os.RemoveAll(tmpMountDir); err2 != nil {
+			err = errors.Append(err, err2)
+		}
+	}()
+
+	if job.RepositoryName != "" {
+		if err := cloneRepo(ctx, tmpMountDir, job, commandRunner, cloneOpts, operations); err != nil {
+			return nil, err
+		}
+	}
+
+	scriptPaths, err := prepareScripts(ctx, job, tmpMountDir, commandRunner, commandLogger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &firecrackerWorkspace{
+		scriptFilenames: scriptPaths,
+		blockDeviceFile: blockDeviceFile,
+		blockDevice:     blockDevice,
+		commandLogger:   commandLogger,
+	}, err
+}
+
+type firecrackerWorkspace struct {
+	scriptFilenames []string
+	blockDeviceFile string
+	blockDevice     string
+	commandLogger   command.Logger
+}
+
+func (w firecrackerWorkspace) Path() string {
+	return w.blockDevice
+}
+
+func (w firecrackerWorkspace) ScriptFilenames() []string {
+	return w.scriptFilenames
+}
+
+func (w firecrackerWorkspace) Remove(ctx context.Context, keepWorkspace bool) {
+	handle := w.commandLogger.Log("teardown.fs", nil)
+	defer func() {
+		// We always finish this with exit code 0 even if it errored, because workspace
+		// cleanup doesn't fail the execution job. We can deal with it separately.
+		handle.Finalize(0)
+		handle.Close()
+	}()
+
+	if keepWorkspace {
+		fmt.Fprintf(handle, "Preserving workspace files (block device: %s, loop file: %s) as per config", w.blockDevice, w.blockDeviceFile)
+		// Remount the workspace, so that it can be inspected.
+		mountDir, err := mountLoopDevice(ctx, w.blockDevice)
+		if err != nil {
+			fmt.Fprintf(handle, "Failed to mount workspace: %s\n", err)
+			return
+		}
+		fmt.Fprintf(handle, "Inspect the workspace contents at: %s\n", mountDir)
```

Comment:
@eseliger dont think we need to do this

```suggestion
		fmt.Fprintf(handle, "Inspect the workspace contents by mounting %q\n", w.blockDevice)
```

Reply:
Why make it harder than necessary? Is there something that'd speak against this?

Reply:
I like that 👍